### PR TITLE
store/tikv: fix region cache data race.

### PR DESCRIPTION
**Pull request by disksing**
RegionCache need update region's fields in multiple threads, we should not use `*Region` without aquire the mutex.

### CODE REVIEW

File: store/tikv/region_cache.go [hunk 1]
```diff
@@ -142,51 +134,58 @@ func (l *KeyLocation) Contains(key []byte) bool {
 
 // LocateKey searches for the region and range that the key is located.
 func (c *RegionCache) LocateKey(bo *Backoffer, key []byte) (*KeyLocation, error) {
-	r := c.getRegionFromCache(key)
+	c.mu.RLock()
+	r := c.searchCachedRegion(key)
 	if r != nil {
```
Comment (siddontang) [inline store/tikv/region_cache.go]: seem you can use RUnlock here

File: store/tikv/region_cache.go [hunk 2]
```diff
@@ -80,35 +86,21 @@ func (c *RPCContext) GetStoreID() uint64 {
...
+	region := c.getCachedRegion(id)
 	if region == nil {
+		c.mu.RUnlock()
 		return nil, nil
 	}
+	// Note: it is safe to use region.meta and region.peer without clone after
+	// unlock, because region cache will never update the content of region's meta
+	// or peer. On the contrary, if we want to use `region` after unlock, then we
+	// need to clone it to avoid data race.
+	meta, peer := region.meta, region.peer
```
Comment (coocood) [inline store/tikv/region_cache.go]: If the `region` never modifies its content, access them with a `region` pointer is safe too.

Comment (disksing) [reply store/tikv/region_cache.go]: `region.peer` may be changed to another peer later, but the content of `reigon.peer` will not be changed.

File: store/tikv/region_cache.go (hunk 1 above)
Comment (disksing) [reply store/tikv/region_cache.go]: I guess not much difference here.

File: store/tikv/region_cache.go [hunk 3]
```diff
@@ -80,35 +86,21 @@ func (c *RPCContext) GetStoreID() uint64 {
...
+	c.mu.RUnlock()
 
 	addr, err := c.GetStoreAddr(bo, region.peer.GetStoreId())
```
Comment (coocood) [inline store/tikv/region_cache.go]: Use `pear` instead of `region.peer`

**Comment by coocood** [comment]
LGTM

**Comment by tiancaiamao** [comment]
LGTM

**Pull request merged**